gitpod-io/gitpod
# jb: java perf monitoring

## Description


See [internal doc](https://www.notion.so/gitpod/Java-Performance-e14d1acd4993410288b95e5687411172) for results of perf testing.

This PR adds tooling which can be used by us or end users to control and monitor resources consumptions of workspaces generally and particularly JB backends:
- adds a supervisor endpoint to fetch resources status (memory and cpu) with respect of cgroup v1
- provide internal CLI command `/.supervisor/supervisor top` to monitor resources status based on endpoint
- contributes env var for each JB product to control max heap size (XMX) via `[product]_XMX`, i.e. `INTELLIJ_XMX=4096M`
- monitor JB backends max/used memory in bytes and report it to prometheus for each workspace pod
- contributes a dev time script `node ./dev/ide/profile-workspace.js` which capture avg/max/90% percentile of memory and CPU usage as well as max/used JB backend memory for a single workspace running on VM

**Out of scope**
- https://github.com/gitpod-io/gitpod/issues/10043
- support of cgroup v2 for resources endpoint [(add a link to new issue here)]
- making it convenient to configure XMX for JB products via .gitpod.yml [[1](https://github.com/gitpod-io/gitpod/issues/8704#issuecomment-1123718534)]
- adding cgroup based stats to JB perf indicator [(add a link to new issue here)]

## Related Issue(s)

Fixes #9521 

## How to test


**Important** only one person can test at the time. So comment below that you are testing and when done remove a comment.

- Start a dev workspace [https://gitpod.io#https://github.com/gitpod-io/gitpod/tree/refs/heads/ak/java_perf](https://gitpod.io/#https://github.com/gitpod-io/gitpod/tree/refs/heads/ak/java_perf)
- Run `./dev/preview/install-k3s-kubeconfig.sh` to configure k8s context.
- Run `./dev/preview/portforward-monitoring-satellite.sh -c harvester` to port forward prometheus API endpoint.
- Run `node ./dev/ide/profile-workspace.js` to start monitoring resource usage of a workspace pod in prev envs.
- Configure prev envs with IntelliJ IDEA: https://ak-java-perf.preview.gitpod-dev.com/preferences
- Start a workspace in prev envs: https://ak-java-perf.preview.gitpod-dev.com/#INTELLIJ_XMX=4096m,JB_MEM_PROFILE=true/https://github.com/eclipse/xtext-core
- Connect with IntelliJ.
- Run `curl http://localhost:22999/_supervisor/v1/status/resources` to check resource status of supervisor.
- Use `/.supervisor/supervisor top` to see resources status of the current workspace via internal CLI
- Capture CPU, memory and heap metrics for the backend from a profile file in the dev workspace: `watch tail -n 1 /workspace/gitpod/dev/ide/perf.log`

## Release Notes

```release-note
NONE
```

## Documentation

- [x] /werft with-vm
- [ ] /werft without-vm
